### PR TITLE
generalize MultiType, add OR and AND subclasses

### PR DESCRIPTION
**Pull request by jakeogh**
Generalize `MultiType` to allow the `OR` and `AND` cases, fix for https://github.com/QCoDeS/Qcodes_contrib_drivers/issues/115

### CODE REVIEW

File: qcodes/tests/validators/test_multi_type.py [hunk 1]
```diff
@@ -21,17 +32,60 @@ def test_good():
 
     assert not MultiType(Strings(), Enum(1, 2)).is_numeric
```
Comment (jenshnielsen) [inline qcodes/tests/validators/test_multi_type.py:34]: I think we might as well split this into two tests e.g. test_good_or and test_good_and

File: qcodes/tests/validators/test_multi_type.py [hunk 2]
```diff
@@ -21,17 +32,60 @@ def test_good():
...
+    #      MultiType(Numbers(min_value=2e-3, max_value=5e4), PermissiveMultiples(divisor=1e-3), combiner='AND'),
+    #      MultiType(Anything(), MultiType(Numbers(), Ints()))]
+    #for m in ms:
+    #    ic(m)
+    #    for val in m.valid_values:
+    #        ic(val)
+    #        m.validate(val)
+
+if __name__ == '__main__':
+    import IPython; IPython.embed()
```
Comment (jenshnielsen) [inline qcodes/tests/validators/test_multi_type.py:91]: What does this do?

File: qcodes/tests/validators/test_multi_type.py [hunk 3]
```diff
@@ -21,17 +32,60 @@ def test_good():
...
+    m = MultiType(Numbers(min_value=2e-3, max_value=5e4), PermissiveMultiples(divisor=1e-3), combiner='AND')
+
+    for v in [10, 11, 123, 1000, 49999.0, 49999.1, 49999.9, 50000.0, 0.1, 0.01, 0.002]:
+        m.validate(v)
+
+    # in py True == 1, so valid in this construction
+    for v in [0, 0.001, 50000.1, 'Q', 'Qcode', None, -1, b'nice', [], {},
+              a_func, AClass, AClass(), False]:
+        with pytest.raises(ValueError):
+            ic(v)
```
Comment (jenshnielsen) [inline qcodes/tests/validators/test_multi_type.py:45]: I think this should be removed ?

Comment (jenshnielsen) [reply qcodes/tests/validators/test_multi_type.py:45]: and a few more places

File: qcodes/utils/validators.py [hunk 4]
```diff
@@ -625,19 +641,111 @@ def validate(self, value: Any, context: str = '') -> None:
...
+                # if ANY of the contained validators is numeric,
+                # the MultiType is considered numeric too.
+                # this could cause problems if you want to sweep
+                # from a non-numeric to a numeric value, so we
+                # need to be careful about this in the sweep code
+                self.is_numeric = True
+
+        self._validators = tuple(validators)
+        self._valid_values = tuple(vval for v in self._validators
+                                   for vval in v._valid_values)
```
Comment (jenshnielsen) [inline qcodes/utils/validators.py:697]: It looks like most of this code is duplicated from the base class and could be removed?

**Comment by jenshnielsen** [comment]
Thanks, Left a few comments inline

File: qcodes/tests/validators/test_multi_type.py (hunk 3 above)
Comment (jakeogh) [reply qcodes/tests/validators/test_multi_type.py]: Fixed.

File: qcodes/tests/validators/test_multi_type.py (hunk 2 above)
Comment (jakeogh) [reply qcodes/tests/validators/test_multi_type.py]: Oops, I left some debugging code in. Fixed.

File: qcodes/tests/validators/test_multi_type.py (hunk 1 above)
Comment (jakeogh) [reply qcodes/tests/validators/test_multi_type.py:32]: Fixed.

File: qcodes/utils/validators.py (hunk 4 above)
Comment (jakeogh) [reply qcodes/utils/validators.py]: Fixed.

File: qcodes/utils/validators.py [hunk 5]
```diff
@@ -625,19 +641,111 @@ def validate(self, value: Any, context: str = '') -> None:
...
```
[the remark this answers is not in the document]
Comment (jakeogh) [reply qcodes/utils/validators.py]: Fixed. Added special case `self._valid_values = ()` for `MultTypeAnd`.

File: qcodes/utils/validators.py [hunk 6]
```diff
@@ -12,6 +12,7 @@
 from typing import Dict as TDict
 from typing import Generic, Hashable
 from typing import List as TList
+from typing import Literal
```
Comment (jenshnielsen) [inline qcodes/utils/validators.py:15]: We need to import Literal from typing_extensions for python 3.7 support until we drop support for 3.7

Comment (jakeogh) [reply qcodes/utils/validators.py]: Fixed!

File: qcodes/tests/validators/test_multi_type.py [hunk 7]
```diff
@@ -16,22 +27,57 @@ def test_good():
...
             m.validate(val)
+
+    # combiner == 'AND'
+    # the concept of checking Validator.valid_values does not hold for the 'AND' case
+    #ms = [MultiType(Numbers(min_value=0, max_value=120), PermissiveMultiples(divisor=0.1), combiner='AND'),
+    #      MultiType(Numbers(min_value=2e-3, max_value=5e4), PermissiveMultiples(divisor=1e-3), combiner='AND'),
+    #      MultiType(Anything(), MultiType(Numbers(), Ints()))]
+    #for m in ms:
+    #    for val in m.valid_values:
+    #        m.validate(val)
```
Comment (jenshnielsen) [inline qcodes/tests/validators/test_multi_type.py:83]: Please just removed this commented out code

File: qcodes/tests/validators/test_multi_type_and.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    assert not MultiTypeAnd(Strings(), Enum(1, 2)).is_numeric
+
+
+def test_bad():
+    for args in [[], [1], [Strings(), True]]:
+        with pytest.raises(TypeError):
+            MultiTypeAnd(*args)
+
+
+def test_valid_values():
```
Comment (jenshnielsen) [inline qcodes/tests/validators/test_multi_type_and.py:30]: Please remove this function

File: qcodes/tests/validators/test_multi_type.py (hunk 7 above)
Comment (jakeogh) [reply qcodes/tests/validators/test_multi_type.py]: done

**Review by jenshnielsen — APPROVED** [review]

**Pull request merged**